Timeouts in session closing and closed reminders

<!-- Before opening a new issue, please check that no existing issue covers the same topic -->
![image](https://cloud.githubusercontent.com/assets/1673303/17759128/843fb1f2-6527-11e6-996a-2d28db03a1a5.png)

One common error I see is this:
```
A problem was encountered with the process that handled this request, causing it to exit. This is likely to cause a new process to be used for the next request to your application. If you see this message frequently, you may be throwing exceptions during the initialization of your application. (Error code 104)
```
Another is this:
```
/feedbackSessionClosedReminders
org.datanucleus.store.appengine.query.QueryExceptionWrappers$JDOQueryTimeoutException
	at org.datanucleus.store.appengine.query.QueryExceptionWrappers$2.get(QueryExceptionWrappers.java:88)
	at org.datanucleus.store.appengine.query.RuntimeExceptionWrappingIterator.hasNext(RuntimeExceptionWrappingIterator.java:105)
	at org.datanucleus.store.appengine.query.LazyResult$LazyAbstractListIterator.computeNext(LazyResult.java:210)
	at org.datanucleus.store.appengine.query.AbstractIterator.tryToComputeNext(AbstractIterator.java:132)
	at org.datanucleus.store.appengine.query.AbstractIterator.hasNext(AbstractIterator.java:127)
	at org.datanucleus.store.appengine.query.LazyResult$AbstractListIterator.hasNext(LazyResult.java:169)
	at teammates.storage.api.FeedbackSessionsDb.getNonPrivateFeedbackSessions(FeedbackSessionsDb.java:162)
	at teammates.logic.core.FeedbackSessionsLogic.getFeedbackSessionsClosedWithinThePastHour(FeedbackSessionsLogic.java:1643)
	at teammates.logic.core.FeedbackSessionsLogic.scheduleFeedbackSessionClosedEmails(FeedbackSessionsLogic.java:1664)
	at teammates.logic.automated.FeedbackSessionClosedRemindersServlet.doGet(FeedbackSessionClosedRemindersServlet.java:24)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:617)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at teammates.storage.datastore.DatastoreFilter.doFilter(DatastoreFilter.java:29)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:125)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.runtime.jetty.SaveSessionFilter.doFilter(SaveSessionFilter.java:37)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:50)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at com.google.apphosting.runtime.jetty.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:260)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:923)
	at com.google.apphosting.runtime.jetty.RpcRequestParser.parseAvailable(RpcRequestParser.java:78)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at com.google.apphosting.runtime.jetty.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:148)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:504)
	at com.google.tracing.TraceContext$TraceContextRunnable.runInContext(TraceContext.java:446)
	at com.google.tracing.TraceContext$TraceContextRunnable$1.run(TraceContext.java:453)
	at com.google.tracing.CurrentContext.runInContext(CurrentContext.java:276)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContextNoUnref(TraceContext.java:312)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContext(TraceContext.java:304)
	at com.google.tracing.TraceContext$TraceContextRunnable.run(TraceContext.java:450)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:235)
	at java.lang.Thread.run(Thread.java:745)
Caused by: org.datanucleus.store.appengine.DatastoreExceptionTranslator$1: The datastore operation timed out, or the data was temporarily unavailable.
	at org.datanucleus.store.appengine.DatastoreExceptionTranslator.wrapDatastoreTimeoutExceptionForQuery(DatastoreExceptionTranslator.java:65)
	at org.datanucleus.store.appengine.query.QueryExceptionWrappers$1.get(QueryExceptionWrappers.java:55)
	at org.datanucleus.store.appengine.query.QueryExceptionWrappers$2.get(QueryExceptionWrappers.java:86)
	... 45 more
Caused by: com.google.appengine.api.datastore.DatastoreTimeoutException: The datastore operation timed out, or the data was temporarily unavailable.
	at com.google.appengine.api.datastore.DatastoreApiHelper.translateError(DatastoreApiHelper.java:64)
	at com.google.appengine.api.datastore.DatastoreApiHelper$1.convertException(DatastoreApiHelper.java:129)
	at com.google.appengine.api.utils.FutureWrapper.get(FutureWrapper.java:97)
	at com.google.appengine.api.datastore.FutureHelper.getInternal(FutureHelper.java:76)
	at com.google.appengine.api.datastore.FutureHelper.quietGet(FutureHelper.java:36)
	at com.google.appengine.api.datastore.BaseQueryResultsSource.loadMoreEntities(BaseQueryResultsSource.java:197)
	at com.google.appengine.api.datastore.BaseQueryResultsSource.loadMoreEntities(BaseQueryResultsSource.java:171)
	at com.google.appengine.api.datastore.QueryResultIteratorImpl.ensureLoaded(QueryResultIteratorImpl.java:148)
	at com.google.appengine.api.datastore.QueryResultIteratorImpl.hasNext(QueryResultIteratorImpl.java:64)
	at com.google.appengine.api.datastore.PreparedMultiQuery$FilteredMultiQueryIterator.computeNext(PreparedMultiQuery.java:249)
	at com.google.appengine.api.datastore.PreparedMultiQuery$FilteredMultiQueryIterator.computeNext(PreparedMultiQuery.java:185)
	at com.google.appengine.api.datastore.AbstractIterator.tryToComputeNext(AbstractIterator.java:46)
	at com.google.appengine.api.datastore.AbstractIterator.hasNext(AbstractIterator.java:41)
	at com.google.appengine.api.datastore.QueryResultIteratorDelegator.hasNext(QueryResultIteratorDelegator.java:38)
	at org.datanucleus.store.appengine.query.RuntimeExceptionWrappingIterator$1.get(RuntimeExceptionWrappingIterator.java:50)
	at org.datanucleus.store.appengine.query.RuntimeExceptionWrappingIterator$1.get(RuntimeExceptionWrappingIterator.java:46)
	at org.datanucleus.store.appengine.query.QueryExceptionWrappers$1.get(QueryExceptionWrappers.java:51)
	... 46 more
```

--- a/src/client/java/teammates/client/scripts/DataMigrationForSentClosingEmailFieldInSessions.java
+++ b/src/client/java/teammates/client/scripts/DataMigrationForSentClosingEmailFieldInSessions.java
@@ -1,0 +1,81 @@
+package teammates.client.scripts;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.jdo.PersistenceManager;
+import javax.jdo.Query;
+
+import teammates.client.remoteapi.RemoteApiClient;
+import teammates.common.datatransfer.FeedbackSessionAttributes;
+import teammates.common.datatransfer.FeedbackSessionType;
+import teammates.common.util.Const.SystemParams;
+import teammates.logic.api.Logic;
+import teammates.storage.datastore.Datastore;
+import teammates.storage.entity.FeedbackSession;
+
+public class DataMigrationForSentClosingEmailFieldInSessions extends RemoteApiClient {
+    
+    private static final Logic logic = new Logic();
+    
+    private boolean isPreview = true;
+    
+    public static void main(String[] args) throws IOException {
+        new DataMigrationForSentClosingEmailFieldInSessions().doOperationRemotely();
+    }
+    
+    @Override
+    protected void doOperation() {
+        Datastore.initialize();
+        
+        List<FeedbackSessionAttributes> sessions = getNonPrivateFeedbackSessions();
+        for (FeedbackSessionAttributes session : sessions) {
+            populateClosingEmailField(session);
+        }
+    }
+    
+    private void populateClosingEmailField(FeedbackSessionAttributes session) {
+        session.setSentClosedEmail(session.isClosed());
+        session.setSentClosingEmail(
+                session.isClosed()
+                || !session.isClosingWithinTimeLimit(SystemParams.NUMBER_OF_HOURS_BEFORE_CLOSING_ALERT));
+        
+        if (isPreview) {
+            System.out.println("sentClosingEmail and sentClosedEmail for " + session.getSessionName()
+                               + " in course " + session.getCourseId() + " to be set to " + session.isClosed());
+            return;
+        }
+        
+        try {
+            logic.updateFeedbackSession(session);
+        } catch (Exception e) {
+            System.out.println("Failed to set attribute sentClosingEmail and sentClosedEmail for session "
+                               + session.getSessionName() + " in course " + session.getCourseId() + ".");
+            e.printStackTrace();
+        }
+    }
+    
+    private List<FeedbackSessionAttributes> getNonPrivateFeedbackSessions() {
+        List<FeedbackSessionAttributes> sessions = new ArrayList<FeedbackSessionAttributes>();
+        List<FeedbackSession> sessionEntities = getNonPrivateFeedbackSessionEntities();
+        for (FeedbackSession sessionEntity : sessionEntities) {
+            sessions.add(new FeedbackSessionAttributes(sessionEntity));
+        }
+        return sessions;
+    }
+    
+    private PersistenceManager getPm() {
+        return Datastore.getPersistenceManager();
+    }
+    
+    @SuppressWarnings("unchecked")
+    private List<FeedbackSession> getNonPrivateFeedbackSessionEntities() {
+        Query q = getPm().newQuery(FeedbackSession.class);
+        q.declareParameters("Enum private");
+        q.setFilter("feedbackSessionType != private");
+        
+        return (List<FeedbackSession>) q.execute(FeedbackSessionType.PRIVATE);
+    }
+    
+}

--- a/src/client/java/teammates/client/scripts/DataMigrationForSentClosingEmailFieldInSessions.java
+++ b/src/client/java/teammates/client/scripts/DataMigrationForSentClosingEmailFieldInSessions.java
@@ -39,7 +39,7 @@ public class DataMigrationForSentClosingEmailFieldInSessions extends RemoteApiCl
         session.setSentClosedEmail(session.isClosed());
         session.setSentClosingEmail(
                 session.isClosed()
-                || !session.isClosingWithinTimeLimit(SystemParams.NUMBER_OF_HOURS_BEFORE_CLOSING_ALERT));
+                || !session.isClosedAfter(SystemParams.NUMBER_OF_HOURS_BEFORE_CLOSING_ALERT));
         
         if (isPreview) {
             System.out.println("sentClosingEmail and sentClosedEmail for " + session.getSessionName()

--- a/src/main/java/teammates/common/datatransfer/FeedbackSessionAttributes.java
+++ b/src/main/java/teammates/common/datatransfer/FeedbackSessionAttributes.java
@@ -33,6 +33,8 @@ public class FeedbackSessionAttributes extends EntityAttributes implements Sessi
     private int gracePeriod;
     private FeedbackSessionType feedbackSessionType;
     private boolean sentOpenEmail;
+    private boolean sentClosingEmail;
+    private boolean sentClosedEmail;
     private boolean sentPublishedEmail;
     private boolean isOpeningEmailEnabled;
     private boolean isClosingEmailEnabled;
@@ -62,6 +64,8 @@ public class FeedbackSessionAttributes extends EntityAttributes implements Sessi
         this.gracePeriod = fs.getGracePeriod();
         this.feedbackSessionType = fs.getFeedbackSessionType();
         this.sentOpenEmail = fs.isSentOpenEmail();
+        this.sentClosingEmail = fs.isSentClosingEmail();
+        this.sentClosedEmail = fs.isSentClosedEmail();
         this.sentPublishedEmail = fs.isSentPublishedEmail();
         this.isOpeningEmailEnabled = fs.isOpeningEmailEnabled();
         this.isClosingEmailEnabled = fs.isClosingEmailEnabled();
@@ -76,20 +80,22 @@ public class FeedbackSessionAttributes extends EntityAttributes implements Sessi
                                      Text instructions, Date createdTime, Date startTime, Date endTime,
                                      Date sessionVisibleFromTime, Date resultsVisibleFromTime,
                                      double timeZone, int gracePeriod, FeedbackSessionType feedbackSessionType,
-                                     boolean sentOpenEmail, boolean sentPublishedEmail,
+                                     boolean sentOpenEmail, boolean sentClosingEmail,
+                                     boolean sentClosedEmail, boolean sentPublishedEmail,
                                      boolean isOpeningEmailEnabled, boolean isClosingEmailEnabled,
                                      boolean isPublishedEmailEnabled) {
         this(feedbackSessionName, courseId, creatorId, instructions, createdTime, startTime, endTime,
              sessionVisibleFromTime, resultsVisibleFromTime, timeZone, gracePeriod, feedbackSessionType,
-             sentOpenEmail, sentPublishedEmail, isOpeningEmailEnabled, isClosingEmailEnabled, isPublishedEmailEnabled,
-             new HashSet<String>(), new HashSet<String>());
+             sentOpenEmail, sentClosingEmail, sentClosedEmail, sentPublishedEmail, isOpeningEmailEnabled,
+             isClosingEmailEnabled, isPublishedEmailEnabled, new HashSet<String>(), new HashSet<String>());
     }
 
     public FeedbackSessionAttributes(String feedbackSessionName, String courseId, String creatorId,
                                      Text instructions, Date createdTime, Date startTime, Date endTime,
                                      Date sessionVisibleFromTime, Date resultsVisibleFromTime,
                                      double timeZone, int gracePeriod, FeedbackSessionType feedbackSessionType,
-                                     boolean sentOpenEmail, boolean sentPublishedEmail,
+                                     boolean sentOpenEmail, boolean sentClosingEmail,
+                                     boolean sentClosedEmail, boolean sentPublishedEmail,
                                      boolean isOpeningEmailEnabled, boolean isClosingEmailEnabled,
                                      boolean isPublishedEmailEnabled, Set<String> instructorList,
                                      Set<String> studentList) {
@@ -107,6 +113,8 @@ public class FeedbackSessionAttributes extends EntityAttributes implements Sessi
         this.gracePeriod = gracePeriod;
         this.feedbackSessionType = feedbackSessionType;
         this.sentOpenEmail = sentOpenEmail;
+        this.sentClosingEmail = sentClosingEmail;
+        this.sentClosedEmail = sentClosedEmail;
         this.sentPublishedEmail = sentPublishedEmail;
         this.isOpeningEmailEnabled = isOpeningEmailEnabled;
         this.isClosingEmailEnabled = isClosingEmailEnabled;
@@ -120,7 +128,7 @@ public class FeedbackSessionAttributes extends EntityAttributes implements Sessi
             other.instructions, other.createdTime, other.startTime, other.endTime,
             other.sessionVisibleFromTime, other.resultsVisibleFromTime, other.timeZone,
             other.gracePeriod, other.feedbackSessionType,
-            other.sentOpenEmail, other.sentPublishedEmail,
+            other.sentOpenEmail, other.sentClosingEmail, other.sentClosedEmail, other.sentPublishedEmail,
             other.isOpeningEmailEnabled, other.isClosingEmailEnabled,
             other.isPublishedEmailEnabled, other.respondingInstructorList,
             other.respondingStudentList);
@@ -154,7 +162,8 @@ public class FeedbackSessionAttributes extends EntityAttributes implements Sessi
     public FeedbackSession toEntity() {
         return new FeedbackSession(feedbackSessionName, courseId, creatorEmail, instructions, createdTime,
                                    startTime, endTime, sessionVisibleFromTime, resultsVisibleFromTime,
-                                   timeZone, gracePeriod, feedbackSessionType, sentOpenEmail, sentPublishedEmail,
+                                   timeZone, gracePeriod, feedbackSessionType, sentOpenEmail,
+                                   sentClosingEmail, sentClosedEmail, sentPublishedEmail,
                                    isOpeningEmailEnabled, isClosingEmailEnabled, isPublishedEmailEnabled,
                                    respondingInstructorList, respondingStudentList);
     }
@@ -639,6 +648,22 @@ public class FeedbackSessionAttributes extends EntityAttributes implements Sessi
 
     public void setSentOpenEmail(boolean sentOpenEmail) {
         this.sentOpenEmail = sentOpenEmail;
+    }
+
+    public boolean isSentClosingEmail() {
+        return sentClosingEmail;
+    }
+
+    public void setSentClosingEmail(boolean sentClosingEmail) {
+        this.sentClosingEmail = sentClosingEmail;
+    }
+
+    public boolean isSentClosedEmail() {
+        return sentClosedEmail;
+    }
+
+    public void setSentClosedEmail(boolean sentClosedEmail) {
+        this.sentClosedEmail = sentClosedEmail;
     }
 
     public boolean isSentPublishedEmail() {

--- a/src/main/java/teammates/common/datatransfer/FeedbackSessionAttributes.java
+++ b/src/main/java/teammates/common/datatransfer/FeedbackSessionAttributes.java
@@ -302,6 +302,24 @@ public class FeedbackSessionAttributes extends EntityAttributes implements Sessi
         return getInvalidityInfo().isEmpty();
     }
 
+    public boolean isClosedAfter(int hours) {
+        Calendar now = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
+        // Fix the time zone accordingly
+        now.add(Calendar.MILLISECOND, (int) (60 * 60 * 1000 * timeZone));
+
+        Calendar start = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
+        start.setTime(startTime);
+
+        Calendar deadline = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
+        deadline.setTime(endTime);
+
+        long nowMillis = now.getTimeInMillis();
+        long deadlineMillis = deadline.getTimeInMillis();
+        long differenceBetweenDeadlineAndNow = (deadlineMillis - nowMillis) / (60 * 60 * 1000);
+
+        return now.after(start) && differenceBetweenDeadlineAndNow < hours;
+    }
+    
     public boolean isClosingWithinTimeLimit(int hours) {
         Calendar now = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
         // Fix the time zone accordingly

--- a/src/main/java/teammates/logic/automated/FeedbackSessionClosedMailAction.java
+++ b/src/main/java/teammates/logic/automated/FeedbackSessionClosedMailAction.java
@@ -6,6 +6,8 @@ import java.util.List;
 import javax.servlet.http.HttpServletRequest;
 
 import teammates.common.datatransfer.FeedbackSessionAttributes;
+import teammates.common.exception.EntityDoesNotExistException;
+import teammates.common.exception.InvalidParametersException;
 import teammates.common.util.Assumption;
 import teammates.common.util.Const;
 import teammates.common.util.Const.ParamsNames;
@@ -42,11 +44,11 @@ public class FeedbackSessionClosedMailAction extends EmailAction {
     }
     
     @Override
-    protected void doPostProcessingForSuccesfulSend() {
-        /* 
-         * Empty because no action is required on successful
-         * sending of feedback session closing mails
-         */
+    protected void doPostProcessingForSuccesfulSend() throws InvalidParametersException, EntityDoesNotExistException {
+        FeedbackSessionAttributes feedbackObject = FeedbackSessionsLogic.inst()
+                .getFeedbackSession(feedbackSessionName, courseId);
+        feedbackObject.setSentClosedEmail(true);
+        FeedbackSessionsLogic.inst().updateFeedbackSession(feedbackObject);
     }
 
     @Override

--- a/src/main/java/teammates/logic/automated/FeedbackSessionClosingMailAction.java
+++ b/src/main/java/teammates/logic/automated/FeedbackSessionClosingMailAction.java
@@ -6,6 +6,8 @@ import java.util.List;
 import javax.servlet.http.HttpServletRequest;
 
 import teammates.common.datatransfer.FeedbackSessionAttributes;
+import teammates.common.exception.EntityDoesNotExistException;
+import teammates.common.exception.InvalidParametersException;
 import teammates.common.util.Assumption;
 import teammates.common.util.Const;
 import teammates.common.util.Const.ParamsNames;
@@ -44,11 +46,11 @@ public class FeedbackSessionClosingMailAction extends EmailAction {
     }
     
     @Override
-    protected void doPostProcessingForSuccesfulSend() {
-        /* 
-         * Empty because no action is required on successful
-         * sending of feedback session closing mails
-         */
+    protected void doPostProcessingForSuccesfulSend() throws InvalidParametersException, EntityDoesNotExistException {
+        FeedbackSessionAttributes feedbackObject = FeedbackSessionsLogic.inst()
+                .getFeedbackSession(feedbackSessionName, courseId);
+        feedbackObject.setSentClosingEmail(true);
+        FeedbackSessionsLogic.inst().updateFeedbackSession(feedbackObject);
     }
 
     @Override

--- a/src/main/java/teammates/logic/core/FeedbackSessionsLogic.java
+++ b/src/main/java/teammates/logic/core/FeedbackSessionsLogic.java
@@ -1622,12 +1622,11 @@ public class FeedbackSessionsLogic {
         ArrayList<FeedbackSessionAttributes> requiredSessions = new
                 ArrayList<FeedbackSessionAttributes>();
 
-        List<FeedbackSessionAttributes> nonPrivateSessions = fsDb
-                .getNonPrivateFeedbackSessions();
+        List<FeedbackSessionAttributes> nonPrivateSessions =
+                fsDb.getFeedbackSessionsNeedingClosingEmail();
 
         for (FeedbackSessionAttributes session : nonPrivateSessions) {
-            if (session.isClosingWithinTimeLimit(SystemParams.NUMBER_OF_HOURS_BEFORE_CLOSING_ALERT)
-                    && session.isClosingEmailEnabled()) {
+            if (session.isClosingWithinTimeLimit(SystemParams.NUMBER_OF_HOURS_BEFORE_CLOSING_ALERT)) {
                 requiredSessions.add(session);
             }
         }
@@ -1640,11 +1639,12 @@ public class FeedbackSessionsLogic {
      */
     public List<FeedbackSessionAttributes> getFeedbackSessionsClosedWithinThePastHour() {
         List<FeedbackSessionAttributes> requiredSessions = new ArrayList<FeedbackSessionAttributes>();
-        List<FeedbackSessionAttributes> nonPrivateSessions = fsDb.getNonPrivateFeedbackSessions();
+        List<FeedbackSessionAttributes> nonPrivateSessions =
+                fsDb.getFeedbackSessionsNeedingClosedEmail();
 
         for (FeedbackSessionAttributes session : nonPrivateSessions) {
-            if (session.isClosedWithinPastHour() // is session closed in the past 1 hour
-                    && session.isClosingEmailEnabled()) {
+            // is session closed in the past 1 hour
+            if (session.isClosedWithinPastHour()) {
                 requiredSessions.add(session);
             }
         }

--- a/src/main/java/teammates/logic/core/FeedbackSessionsLogic.java
+++ b/src/main/java/teammates/logic/core/FeedbackSessionsLogic.java
@@ -2639,22 +2639,27 @@ public class FeedbackSessionsLogic {
     private void makeEmailStateConsistent(FeedbackSessionAttributes oldSession,
             FeedbackSessionAttributes newSession) {
 
-        // reset sentOpenEmail if the session has opened but is being closed
-        // now.
-        if (oldSession.isSentOpenEmail() && !newSession.isOpened()) {
-            newSession.setSentOpenEmail(false);
-        } else if (oldSession.isSentOpenEmail()) {
-            // or else leave it as sent if so.
-            newSession.setSentOpenEmail(true);
+        // reset sentOpenEmail if the session has opened but is being un-opened
+        // now, or else leave it as sent if so.
+        if (oldSession.isSentOpenEmail()) {
+            newSession.setSentOpenEmail(newSession.isOpened());
+        }
+
+        // reset sentClosedEmail if the session has closed but is being un-closed
+        // now, or else leave it as sent if so.
+        if (oldSession.isSentClosedEmail()) {
+            newSession.setSentClosedEmail(newSession.isClosed());
+            
+            // also reset sentClosingEmail
+            newSession.setSentClosingEmail(
+                    newSession.isClosed()
+                    || !newSession.isClosingWithinTimeLimit(SystemParams.NUMBER_OF_HOURS_BEFORE_CLOSING_ALERT));
         }
 
         // reset sentPublishedEmail if the session has been published but is
-        // going to be unpublished now.
-        if (oldSession.isSentPublishedEmail() && !newSession.isPublished()) {
-            newSession.setSentPublishedEmail(false);
-        } else if (oldSession.isSentPublishedEmail()) {
-            // or else leave it as sent if so.
-            newSession.setSentPublishedEmail(true);
+        // going to be unpublished now, or else leave it as sent if so.
+        if (oldSession.isSentPublishedEmail()) {
+            newSession.setSentPublishedEmail(newSession.isPublished());
         }
     }
 

--- a/src/main/java/teammates/logic/core/FeedbackSessionsLogic.java
+++ b/src/main/java/teammates/logic/core/FeedbackSessionsLogic.java
@@ -2653,7 +2653,7 @@ public class FeedbackSessionsLogic {
             // also reset sentClosingEmail
             newSession.setSentClosingEmail(
                     newSession.isClosed()
-                    || !newSession.isClosingWithinTimeLimit(SystemParams.NUMBER_OF_HOURS_BEFORE_CLOSING_ALERT));
+                    || !newSession.isClosedAfter(SystemParams.NUMBER_OF_HOURS_BEFORE_CLOSING_ALERT));
         }
 
         // reset sentPublishedEmail if the session has been published but is

--- a/src/main/java/teammates/storage/api/FeedbackSessionsDb.java
+++ b/src/main/java/teammates/storage/api/FeedbackSessionsDb.java
@@ -564,10 +564,10 @@ public class FeedbackSessionsDb extends EntitiesDb {
     private List<FeedbackSession> getFeedbackSessionEntitiesNeedingClosedEmail() {
         Query q = getPm().newQuery(FeedbackSession.class);
         q.declareParameters("boolean sentParam, boolean enableParam, Enum notTypeParam");
-        q.setFilter("sentClosedEmail == sentParam && isClosedEmailEnabled == enableParam "
+        q.setFilter("sentClosedEmail == sentParam && isClosingEmailEnabled == enableParam "
                     + "&& feedbackSessionType != notTypeParam");
         
-        return (List<FeedbackSession>) q.execute(false, false, FeedbackSessionType.PRIVATE);
+        return (List<FeedbackSession>) q.execute(false, true, FeedbackSessionType.PRIVATE);
     }
     
     @SuppressWarnings("unchecked")

--- a/src/main/java/teammates/storage/api/FeedbackSessionsDb.java
+++ b/src/main/java/teammates/storage/api/FeedbackSessionsDb.java
@@ -152,24 +152,6 @@ public class FeedbackSessionsDb extends EntitiesDb {
     /**
      * Preconditions: <br>
      * * All parameters are non-null.
-     * @return An empty list if no non-private sessions are found.
-     */
-    public List<FeedbackSessionAttributes> getNonPrivateFeedbackSessions() {
-        
-        List<FeedbackSession> fsList = getNonPrivateFeedbackSessionEntities();
-        List<FeedbackSessionAttributes> fsaList = new ArrayList<FeedbackSessionAttributes>();
-        
-        for (FeedbackSession fs : fsList) {
-            if (!JDOHelper.isDeleted(fs)) {
-                fsaList.add(new FeedbackSessionAttributes(fs));
-            }
-        }
-        return fsaList;
-    }
-        
-    /**
-     * Preconditions: <br>
-     * * All parameters are non-null.
      * @return An empty list if no sessions are found for the given course.
      */
     public List<FeedbackSessionAttributes> getFeedbackSessionsForCourse(String courseId) {
@@ -186,7 +168,7 @@ public class FeedbackSessionsDb extends EntitiesDb {
         }
         return fsaList;
     }
-    
+        
     /**
      * Preconditions: <br>
      * * All parameters are non-null.
@@ -195,6 +177,38 @@ public class FeedbackSessionsDb extends EntitiesDb {
     public List<FeedbackSessionAttributes> getFeedbackSessionsWithUnsentOpenEmail() {
                 
         List<FeedbackSession> fsList = getFeedbackSessionEntitiesWithUnsentOpenEmail();
+        List<FeedbackSessionAttributes> fsaList = new ArrayList<FeedbackSessionAttributes>();
+        
+        for (FeedbackSession fs : fsList) {
+            if (!JDOHelper.isDeleted(fs)) {
+                fsaList.add(new FeedbackSessionAttributes(fs));
+            }
+        }
+        return fsaList;
+    }
+    
+    /**
+     * @return An empty list if no sessions are found that have unsent closing emails.
+     */
+    public List<FeedbackSessionAttributes> getFeedbackSessionsNeedingClosingEmail() {
+                
+        List<FeedbackSession> fsList = getFeedbackSessionEntitiesNeedingClosingEmail();
+        List<FeedbackSessionAttributes> fsaList = new ArrayList<FeedbackSessionAttributes>();
+        
+        for (FeedbackSession fs : fsList) {
+            if (!JDOHelper.isDeleted(fs)) {
+                fsaList.add(new FeedbackSessionAttributes(fs));
+            }
+        }
+        return fsaList;
+    }
+    
+    /**
+     * @return An empty list if no sessions are found that have unsent closed emails.
+     */
+    public List<FeedbackSessionAttributes> getFeedbackSessionsNeedingClosedEmail() {
+                
+        List<FeedbackSession> fsList = getFeedbackSessionEntitiesNeedingClosedEmail();
         List<FeedbackSessionAttributes> fsaList = new ArrayList<FeedbackSessionAttributes>();
         
         for (FeedbackSession fs : fsList) {
@@ -261,6 +275,8 @@ public class FeedbackSessionsDb extends EntitiesDb {
         fs.setGracePeriod(newAttributes.getGracePeriod());
         fs.setFeedbackSessionType(newAttributes.getFeedbackSessionType());
         fs.setSentOpenEmail(newAttributes.isSentOpenEmail());
+        fs.setSentClosingEmail(newAttributes.isSentClosingEmail());
+        fs.setSentClosedEmail(newAttributes.isSentClosedEmail());
         fs.setSentPublishedEmail(newAttributes.isSentPublishedEmail());
         fs.setIsOpeningEmailEnabled(newAttributes.isOpeningEmailEnabled());
         fs.setSendClosingEmail(newAttributes.isClosingEmailEnabled());
@@ -517,15 +533,6 @@ public class FeedbackSessionsDb extends EntitiesDb {
     }
     
     @SuppressWarnings("unchecked")
-    private List<FeedbackSession> getNonPrivateFeedbackSessionEntities() {
-        Query q = getPm().newQuery(FeedbackSession.class);
-        q.declareParameters("Enum private");
-        q.setFilter("feedbackSessionType != private");
-        
-        return (List<FeedbackSession>) q.execute(FeedbackSessionType.PRIVATE);
-    }
-    
-    @SuppressWarnings("unchecked")
     private List<FeedbackSession> getFeedbackSessionEntitiesForCourse(String courseId) {
         Query q = getPm().newQuery(FeedbackSession.class);
         q.declareParameters("String courseIdParam");
@@ -541,6 +548,26 @@ public class FeedbackSessionsDb extends EntitiesDb {
         q.setFilter("sentOpenEmail == sentParam && feedbackSessionType != notTypeParam");
         
         return (List<FeedbackSession>) q.execute(false, FeedbackSessionType.PRIVATE);
+    }
+    
+    @SuppressWarnings("unchecked")
+    private List<FeedbackSession> getFeedbackSessionEntitiesNeedingClosingEmail() {
+        Query q = getPm().newQuery(FeedbackSession.class);
+        q.declareParameters("boolean sentParam, boolean enableParam, Enum notTypeParam");
+        q.setFilter("sentClosingEmail == sentParam && isClosingEmailEnabled == enableParam "
+                    + "&& feedbackSessionType != notTypeParam");
+        
+        return (List<FeedbackSession>) q.execute(false, true, FeedbackSessionType.PRIVATE);
+    }
+    
+    @SuppressWarnings("unchecked")
+    private List<FeedbackSession> getFeedbackSessionEntitiesNeedingClosedEmail() {
+        Query q = getPm().newQuery(FeedbackSession.class);
+        q.declareParameters("boolean sentParam, boolean enableParam, Enum notTypeParam");
+        q.setFilter("sentClosedEmail == sentParam && isClosedEmailEnabled == enableParam "
+                    + "&& feedbackSessionType != notTypeParam");
+        
+        return (List<FeedbackSession>) q.execute(false, false, FeedbackSessionType.PRIVATE);
     }
     
     @SuppressWarnings("unchecked")

--- a/src/main/java/teammates/storage/entity/FeedbackSession.java
+++ b/src/main/java/teammates/storage/entity/FeedbackSession.java
@@ -94,6 +94,12 @@ public class FeedbackSession {
     private boolean sentOpenEmail;
     
     @Persistent
+    private Boolean sentClosingEmail;
+    
+    @Persistent
+    private Boolean sentClosedEmail;
+    
+    @Persistent
     private boolean sentPublishedEmail;
 
     //TODO change to primitive types and update getter
@@ -109,18 +115,20 @@ public class FeedbackSession {
     public FeedbackSession(String feedbackSessionName, String courseId,
             String creatorEmail, Text instructions, Date createdTime, Date startTime, Date endTime,
             Date sessionVisibleFromTime, Date resultsVisibleFromTime, double timeZone, int gracePeriod,
-            FeedbackSessionType feedbackSessionType, boolean sentOpenEmail, boolean sentPublishedEmail,
+            FeedbackSessionType feedbackSessionType, boolean sentOpenEmail,
+            boolean sentClosingEmail, boolean sentClosedEmail, boolean sentPublishedEmail,
             boolean isOpeningEmailEnabled, boolean isClosingEmailEnabled, boolean isPublishedEmailEnabled) {
         this(feedbackSessionName, courseId, creatorEmail, instructions, createdTime, startTime, endTime,
-             sessionVisibleFromTime, resultsVisibleFromTime, timeZone, gracePeriod,
-             feedbackSessionType, sentOpenEmail, sentPublishedEmail, isOpeningEmailEnabled,
+             sessionVisibleFromTime, resultsVisibleFromTime, timeZone, gracePeriod, feedbackSessionType,
+             sentOpenEmail, sentClosingEmail, sentClosedEmail, sentPublishedEmail, isOpeningEmailEnabled,
              isClosingEmailEnabled, isPublishedEmailEnabled, new HashSet<String>(), new HashSet<String>());
     }
 
     public FeedbackSession(String feedbackSessionName, String courseId,
             String creatorEmail, Text instructions, Date createdTime, Date startTime, Date endTime,
             Date sessionVisibleFromTime, Date resultsVisibleFromTime, double timeZone, int gracePeriod,
-            FeedbackSessionType feedbackSessionType, boolean sentOpenEmail, boolean sentPublishedEmail,
+            FeedbackSessionType feedbackSessionType, boolean sentOpenEmail, boolean sentClosingEmail,
+            boolean sentClosedEmail, boolean sentPublishedEmail,
             boolean isOpeningEmailEnabled, boolean isClosingEmailEnabled, boolean isPublishedEmailEnabled,
             Set<String> instructorList, Set<String> studentList) {
         this.feedbackSessionName = feedbackSessionName;
@@ -137,6 +145,8 @@ public class FeedbackSession {
         this.gracePeriod = gracePeriod;
         this.feedbackSessionType = feedbackSessionType;
         this.sentOpenEmail = sentOpenEmail;
+        this.sentClosingEmail = sentClosingEmail;
+        this.sentClosedEmail = sentClosedEmail;
         this.sentPublishedEmail = sentPublishedEmail;
         this.isOpeningEmailEnabled = isOpeningEmailEnabled;
         this.isClosingEmailEnabled = isClosingEmailEnabled;
@@ -257,8 +267,32 @@ public class FeedbackSession {
         return sentOpenEmail;
     }
 
-    public void setSentOpenEmail(boolean activated) {
-        this.sentOpenEmail = activated;
+    public void setSentOpenEmail(boolean sentOpenEmail) {
+        this.sentOpenEmail = sentOpenEmail;
+    }
+
+    public boolean isSentClosingEmail() {
+        // Legacy data might not have this field
+        if (sentClosingEmail == null) {
+            return false;
+        }
+        return sentClosedEmail;
+    }
+
+    public void setSentClosingEmail(boolean sentClosingEmail) {
+        this.sentClosingEmail = sentClosingEmail;
+    }
+
+    public boolean isSentClosedEmail() {
+        // Legacy data might not have this field
+        if (sentClosedEmail == null) {
+            return false;
+        }
+        return sentClosedEmail;
+    }
+
+    public void setSentClosedEmail(boolean sentClosedEmail) {
+        this.sentClosedEmail = sentClosedEmail;
     }
 
     public boolean isSentPublishedEmail() {

--- a/src/test/java/teammates/test/cases/common/FeedbackSessionAttributesTest.java
+++ b/src/test/java/teammates/test/cases/common/FeedbackSessionAttributesTest.java
@@ -30,6 +30,8 @@ public class FeedbackSessionAttributesTest extends BaseTestCase {
     private static int gracePeriod;
     private static FeedbackSessionType feedbackSessionType;
     private static boolean sentOpenEmail;
+    private static boolean sentClosingEmail;
+    private static boolean sentClosedEmail;
     private static boolean sentPublishedEmail;
     private static boolean isOpeningEmailEnabled;
     private static boolean isClosingEmailEnabled;
@@ -54,6 +56,8 @@ public class FeedbackSessionAttributesTest extends BaseTestCase {
         gracePeriod = 15;
         feedbackSessionType = FeedbackSessionType.STANDARD;
         sentOpenEmail = false;
+        sentClosingEmail = false;
+        sentClosedEmail = false;
         sentPublishedEmail = false;
         isOpeningEmailEnabled = false;
         isClosingEmailEnabled = false;
@@ -63,8 +67,8 @@ public class FeedbackSessionAttributesTest extends BaseTestCase {
                 courseId, creatorId, instructions,
                 createdTime, startTime, endTime,
                 sessionVisibleFromTime, resultsVisibleFromTime,
-                timeZone, gracePeriod,
-                feedbackSessionType, sentOpenEmail, sentPublishedEmail,
+                timeZone, gracePeriod, feedbackSessionType,
+                sentOpenEmail, sentClosingEmail, sentClosedEmail, sentPublishedEmail,
                 isOpeningEmailEnabled, isClosingEmailEnabled, isPublishedEmailEnabled
                 );
     }

--- a/src/test/java/teammates/test/cases/logic/StudentsLogicTest.java
+++ b/src/test/java/teammates/test/cases/logic/StudentsLogicTest.java
@@ -648,7 +648,7 @@ public class StudentsLogicTest extends BaseComponentTestCase {
                 TimeHelper.getHoursOffsetToCurrentTime(0), TimeHelper.getHoursOffsetToCurrentTime(2),
                 TimeHelper.getHoursOffsetToCurrentTime(5), TimeHelper.getHoursOffsetToCurrentTime(1),
                 TimeHelper.getHoursOffsetToCurrentTime(6),
-                8.0, 0, FeedbackSessionType.PRIVATE, false, false, false, false, false);
+                8.0, 0, FeedbackSessionType.PRIVATE, false, false, false, false, false, false, false);
         fsLogic.createFeedbackSession(fsAttr);
         
         

--- a/src/test/java/teammates/test/cases/storage/FeedbackSessionsDbTest.java
+++ b/src/test/java/teammates/test/cases/storage/FeedbackSessionsDbTest.java
@@ -100,7 +100,6 @@ public class FeedbackSessionsDbTest extends BaseComponentTestCase {
 
         testGetFeedbackSessions();
         testGetFeedbackSessionsForCourse();
-        testGetNonPrivateFeedbackSessions();
         testGetFeedbackSessionsWithUnsentOpenEmail();
         testGetFeedbackSessionsWithUnsentPublishedEmail();
     }
@@ -175,19 +174,6 @@ public class FeedbackSessionsDbTest extends BaseComponentTestCase {
         ______TS("no sessions in course");
         
         assertTrue(fsDb.getFeedbackSessionsForCourse("idOfCourseNoEvals").isEmpty());
-    }
-    
-    private void testGetNonPrivateFeedbackSessions() {
-        
-        ______TS("standard success case");
-        
-        List<FeedbackSessionAttributes> fsaList = fsDb.getNonPrivateFeedbackSessions();
-        
-        assertEquals(8, fsaList.size());
-        for (FeedbackSessionAttributes fsa : fsaList) {
-            assertFalse(fsa.isPrivateSession());
-        }
-        
     }
     
     private void testGetFeedbackSessionsWithUnsentOpenEmail() {

